### ISSUE TITLE
[BUG] fail to update all global installed packages

### ISSUE THREAD

**Issue by CyrusYip** [npm/cli#2415]
### Current Behavior:

I used `sudo npm-check --global --update--all` and `sudo npm update -g` to update global installed packages, but the terminal showed `npm ERR! Maximum call stack size exceeded`. However, I can globally install a single package.

```bash
sudo npm install --global @marp-team/marp-cli@0.23.0 --color=always
[sudo] password for hunter: 
npm WARN deprecated fsevents@2.1.3: Please update to v 2.2.x
/usr/local/bin/marp -> /usr/local/lib/node_modules/@marp-team/marp-cli/marp-cli.js
npm WARN optional SKIPPING OPTIONAL DEPENDENCY: fsevents@~2.1.2 (node_modules/@marp-team/marp-cli/node_modules/chokidar/node_modules/fsevents):
npm WARN notsup SKIPPING OPTIONAL DEPENDENCY: Unsupported platform for fsevents@2.1.3: wanted {"os":"darwin","arch":"any"} (current: {"os":"linux","arch":"x64"})

+ @marp-team/marp-cli@0.23.0
removed 7 packages and updated 1 package in 41.186s
```

### Expected Behavior:

I can use `sudo npm-check --global --update--all` and `sudo npm update -g` to update global installed packages.

### Steps To Reproduce:

This problem can be reproduce whenever I execute `sudo npm update -g` and `sudo npm-check --global --update--all`.

```bash
$ sudo npm update -g
npm ERR! Maximum call stack size exceeded

npm ERR! A complete log of this run can be found in:
npm ERR!     /root/.npm/_logs/2020-12-26T00_29_31_080Z-debug.log
```

<details><summary>2020-12-26T00_29_31_080Z-debug.log</summary>

```
0 info it worked if it ends with ok
1 verbose cli [ '/usr/bin/node', '/usr/local/bin/npm', 'update', '-g' ]
2 info using npm@6.14.10
3 info using node@v12.18.2
4 verbose npm-session f40621286aac5df0
5 verbose update computing outdated modules to update
6 http fetch GET 304 https://registry.npmjs.org/npm-check 1575ms (from cache)
7 silly pacote registry packument for npm-check@^5.9.2 fetched in 1604ms
8 http fetch GET 304 https://registry.npmjs.org/docsify-cli 1589ms (from cache)
9 http fetch GET 304 https://registry.npmjs.org/markmap-lib 1589ms (from cache)
10 http fetch GET 304 https://registry.npmjs.org/markmap-cli 1592ms (from cache)
11 silly pacote registry packument for docsify-cli@^4.4.2 fetched in 1615ms
12 silly pacote registry packument for markmap-lib@^0.8.0 fetched in 1614ms
13 silly pacote registry packument for markmap-cli@^0.4.4 fetched in 1615ms
14 http fetch GET 304 https://registry.npmjs.org/yarn 1589ms (from cache)
15 silly pacote registry packument for yarn@^1.22.10 fetched in 1621ms
16 http fetch GET 304 https://registry.npmjs.org/netlify-cli 1610ms (from cache)
17 silly pacote registry packument for netlify-cli@^2.65.6 fetched in 1649ms
18 http fetch GET 304 https://registry.npmjs.org/npm 1630ms (from cache)
19 silly pacote registry packument for npm@^6.14.10 fetched in 1684ms
20 http fetch GET 304 https://registry.npmjs.org/npm-check-updates 2083ms (from cache)
21 silly pacote registry packument for npm-check-updates@^10.2.5 fetched in 2115ms
22 http fetch GET 304 https://registry.npmjs.org/cnpm 2711ms (from cache)
23 silly pacote registry packument for cnpm@^6.1.1 fetched in 2732ms
24 http fetch GET 304 https://registry.npmjs.org/@marp-team%2fmarp-cli 2722ms (from cache)
25 silly pacote registry packument for @marp-team/marp-cli@^0.23.0 fetched in 2745ms
26 info outdated updating [
26 info outdated   {
26 info outdated     dep: Node {
26 info outdated       name: 'markmap-lib',
26 info outdated       path: '/usr/local/lib/node_modules/markmap-lib',
26 info outdated       realpath: '/usr/local/lib/node_modules/markmap-lib',
26 info outdated       error: null,
26 info outdated       id: 1,
26 info outdated       package: [Object],
26 info outdated       parent: [Node],
26 info outdated       isLink: false,
26 info outdated       children: [Array],
26 info outdated       loaded: false,
26 info outdated       requiredBy: [],
26 info outdated       requires: [Array],
26 info outdated       missingDeps: {},
26 info outdated       missingDevDeps: {},
26 info outdated       phantomChildren: [Object],
26 info outdated       location: '/markmap-lib',
26 info outdated       userRequired: false,
26 info outdated       save: false,
26 info outdated       saveSpec: null,
26 info outdated       isTop: false,
26 info outdated       fromBundle: false,
26 info outdated       isInLink: false
26 info outdated     },
26 info outdated     depname: 'markmap-lib',
26 info outdated     current: '0.8.0',
26 info outdated     wanted: '0.8.1',
26 info outdated     latest: '0.11.1',
26 info outdated     req: '^0.8.0',
26 info outdated     what: 'markmap-lib@0.8.1'
26 info outdated   },
26 info outdated   {
26 info outdated     dep: Node {
26 info outdated       name: 'netlify-cli',
26 info outdated       path: '/usr/local/lib/node_modules/netlify-cli',
26 info outdated       realpath: '/usr/local/lib/node_modules/netlify-cli',
26 info outdated       error: null,
26 info outdated       id: 2,
26 info outdated       package: [Object],
26 info outdated       parent: [Node],
26 info outdated       isLink: false,
26 info outdated       children: [Array],
26 info outdated       loaded: false,
26 info outdated       requiredBy: [],
26 info outdated       requires: [Array],
26 info outdated       missingDeps: {},
26 info outdated       missingDevDeps: {},
26 info outdated       phantomChildren: [Object],
26 info outdated       location: '/netlify-cli',
26 info outdated       userRequired: false,
26 info outdated       save: false,
26 info outdated       saveSpec: null,
26 info outdated       isTop: false,
26 info outdated       fromBundle: false,
26 info outdated       isInLink: false
26 info outdated     },
26 info outdated     depname: 'netlify-cli',
26 info outdated     current: '2.65.6',
26 info outdated     wanted: '2.69.11',
26 info outdated     latest: '2.69.11',
26 info outdated     req: '^2.65.6',
26 info outdated     what: 'netlify-cli@2.69.11'
26 info outdated   }
26 info outdated ]
27 silly install loadCurrentTree
28 silly install readGlobalPackageData
29 http fetch GET 304 https://registry.npmjs.org/netlify-cli 277ms (from cache)
30 http fetch GET 304 https://registry.npmjs.org/markmap-lib 287ms (from cache)
31 silly pacote version manifest for netlify-cli@2.69.11 fetched in 289ms
32 silly pacote version manifest for markmap-lib@0.8.1 fetched in 293ms
33 timing stage:loadCurrentTree Completed in 7059ms
34 silly install loadIdealTree
35 silly install cloneCurrentTreeToIdealTree
36 timing stage:loadIdealTree:cloneCurrentTree Completed in 90ms
37 silly install loadShrinkwrap
38 verbose stack RangeError: Maximum call stack size exceeded
38 verbose stack     at /usr/local/lib/node_modules/npm/node_modules/aproba/index.js:39:16
38 verbose stack     at Array.forEach (<anonymous>)
38 verbose stack     at validate (/usr/local/lib/node_modules/npm/node_modules/aproba/index.js:34:11)
38 verbose stack     at flatNameFromTree (/usr/local/lib/node_modules/npm/lib/install/flatten-tree.js:37:3)
38 verbose stack     at flatNameFromTree (/usr/local/lib/node_modules/npm/lib/install/flatten-tree.js:39:14)
38 verbose stack     at flatNameFromTree (/usr/local/lib/node_modules/npm/lib/install/flatten-tree.js:39:14)
38 verbose stack     at flatNameFromTree (/usr/local/lib/node_modules/npm/lib/install/flatten-tree.js:39:14)
38 verbose stack     at flatNameFromTree (/usr/local/lib/node_modules/npm/lib/install/flatten-tree.js:39:14)
38 verbose stack     at flatNameFromTree (/usr/local/lib/node_modules/npm/lib/install/flatten-tree.js:39:14)
38 verbose stack     at flatNameFromTree (/usr/local/lib/node_modules/npm/lib/install/flatten-tree.js:39:14)
38 verbose stack     at flatNameFromTree (/usr/local/lib/node_modules/npm/lib/install/flatten-tree.js:39:14)
38 verbose stack     at flatNameFromTree (/usr/local/lib/node_modules/npm/lib/install/flatten-tree.js:39:14)
38 verbose stack     at flatNameFromTree (/usr/local/lib/node_modules/npm/lib/install/flatten-tree.js:39:14)
38 verbose stack     at flatNameFromTree (/usr/local/lib/node_modules/npm/lib/install/flatten-tree.js:39:14)
38 verbose stack     at flatNameFromTree (/usr/local/lib/node_modules/npm/lib/install/flatten-tree.js:39:14)
38 verbose stack     at flatNameFromTree (/usr/local/lib/node_modules/npm/lib/install/flatten-tree.js:39:14)
39 verbose cwd /home/hunter/Documents/my_git_repo/cyrusyip-blog
40 verbose Linux 5.8.0-33-generic
41 verbose argv "/usr/bin/node" "/usr/local/bin/npm" "update" "-g"
42 verbose node v12.18.2
43 verbose npm  v6.14.10
44 error Maximum call stack size exceeded
45 verbose exit [ 1, true ]

```
</details>

```bash
$ sudo npm-check --global --update--all                              
The global path you are searching is: /usr/local/lib/node_modules

$ npm install --global markmap-lib@0.11.1 netlify-cli@2.69.11 --color=always
Command failed: npm install --global markmap-lib@0.11.1 netlify-cli@2.69.11 --color=always
npm ERR! Maximum call stack size exceeded

npm ERR! A complete log of this run can be found in:
npm ERR!     /root/.npm/_logs/2020-12-26T00_22_54_104Z-debug.log

For more detail, add `--debug` to the command
```

<details><summary>2020-12-26T00_22_54_104Z-debug.log</summary>

```
0 info it worked if it ends with ok
1 verbose cli [
1 verbose cli   '/usr/bin/node',
1 verbose cli   '/usr/bin/npm',
1 verbose cli   'install',
1 verbose cli   '--global',
1 verbose cli   'markmap-lib@0.11.1',
1 verbose cli   'netlify-cli@2.69.11',
1 verbose cli   '--color=always'
1 verbose cli ]
2 info using npm@6.14.8
3 info using node@v12.18.2
4 verbose npm-session dffb19f6d04b488d
5 silly install loadCurrentTree
6 silly install readGlobalPackageData
7 http fetch GET 304 https://registry.npmjs.org/netlify-cli 1154ms (from cache)
8 silly pacote version manifest for netlify-cli@2.69.11 fetched in 1178ms
9 http fetch GET 304 https://registry.npmjs.org/markmap-lib 1281ms (from cache)
10 silly pacote version manifest for markmap-lib@0.11.1 fetched in 1304ms
11 timing stage:loadCurrentTree Completed in 9086ms
12 silly install loadIdealTree
13 silly install cloneCurrentTreeToIdealTree
14 timing stage:loadIdealTree:cloneCurrentTree Completed in 97ms
15 silly install loadShrinkwrap
16 verbose stack RangeError: Maximum call stack size exceeded
16 verbose stack     at /usr/share/nodejs/aproba/index.js:39:16
16 verbose stack     at Array.forEach (<anonymous>)
16 verbose stack     at validate (/usr/share/nodejs/aproba/index.js:34:11)
16 verbose stack     at flatNameFromTree (/usr/share/npm/lib/install/flatten-tree.js:37:3)
16 verbose stack     at flatNameFromTree (/usr/share/npm/lib/install/flatten-tree.js:39:14)
16 verbose stack     at flatNameFromTree (/usr/share/npm/lib/install/flatten-tree.js:39:14)
16 verbose stack     at flatNameFromTree (/usr/share/npm/lib/install/flatten-tree.js:39:14)
16 verbose stack     at flatNameFromTree (/usr/share/npm/lib/install/flatten-tree.js:39:14)
16 verbose stack     at flatNameFromTree (/usr/share/npm/lib/install/flatten-tree.js:39:14)
16 verbose stack     at flatNameFromTree (/usr/share/npm/lib/install/flatten-tree.js:39:14)
16 verbose stack     at flatNameFromTree (/usr/share/npm/lib/install/flatten-tree.js:39:14)
16 verbose stack     at flatNameFromTree (/usr/share/npm/lib/install/flatten-tree.js:39:14)
16 verbose stack     at flatNameFromTree (/usr/share/npm/lib/install/flatten-tree.js:39:14)
16 verbose stack     at flatNameFromTree (/usr/share/npm/lib/install/flatten-tree.js:39:14)
16 verbose stack     at flatNameFromTree (/usr/share/npm/lib/install/flatten-tree.js:39:14)
16 verbose stack     at flatNameFromTree (/usr/share/npm/lib/install/flatten-tree.js:39:14)
17 verbose cwd /usr/local/lib/node_modules
18 verbose Linux 5.8.0-33-generic
19 verbose argv "/usr/bin/node" "/usr/bin/npm" "install" "--global" "markmap-lib@0.11.1" "netlify-cli@2.69.11" "--color=always"
20 verbose node v12.18.2
21 verbose npm  v6.14.8
22 error Maximum call stack size exceeded
23 verbose exit [ 1, true ]

```
</details>

### Environment:

- Kubuntu 20.10
- Node: v12.18.2
- npm: 6.14.10

```bash
$ lsb_release -a
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 20.10
Release:        20.10
Codename:       groovy

$ node -v
v12.18.2

$ npm -v           
6.14.10
```

**Comment by ljharb** [comment]
What happens with npm v7.3.0?

**Comment by CyrusYip** [comment]
npm v7.3.0 did not work.

```
~ sudo npm install -g npm@next-7

/usr/local/bin/npm -> /usr/local/lib/node_modules/npm/bin/npm-cli.js
/usr/local/bin/npx -> /usr/local/lib/node_modules/npm/bin/npx-cli.js
+ npm@7.3.0
added 56 packages from 28 contributors, removed 241 packages and updated 194 packages in 19.551s
~ npm -v
7.3.0

~ npm-check --global              
The global path you are searching is: /usr/local/lib/node_modules

markmap-lib   😎  NEW VER!  NonSemver update available. https://github.com/gera2ld/markmap#readme
                           npm install --global markmap-lib@0.11.1 to go from 0.8.0 to 0.11.1

netlify-cli   😎  MINOR UP  Minor update available. https://github.com/netlify/cli
                           npm install --global netlify-cli@2.69.11 to go from 2.65.6 to 2.69.11

Use npm-check -gu for interactive update.

~ sudo npm-check --global --update--all
The global path you are searching is: /usr/local/lib/node_modules

$ npm install --global markmap-lib@0.11.1 netlify-cli@2.69.11 --color=always
Command failed: npm install --global markmap-lib@0.11.1 netlify-cli@2.69.11 --color=always
npm ERR! Maximum call stack size exceeded

npm ERR! A complete log of this run can be found in:
npm ERR!     /root/.npm/_logs/2020-12-26T06_21_47_592Z-debug.log

For more detail, add `--debug` to the command

~ npm outdated -g --depth=0

Package      Current   Wanted   Latest  Location                  Depended by
markmap-lib    0.8.0   0.11.1   0.11.1  node_modules/markmap-lib  global
netlify-cli   2.65.6  2.69.11  2.69.11  node_modules/netlify-cli  global
npm            7.3.0  6.14.10  6.14.10  node_modules/npm          global
~ sudo npm update -g                                                   

up to date, audited 1 package in 263ms

found 0 vulnerabilities

```

<details><summary>2020-12-26T06_21_47_592Z-debug</summary>

```
0 info it worked if it ends with ok
1 verbose cli [
1 verbose cli   '/usr/bin/node',
1 verbose cli   '/usr/bin/npm',
1 verbose cli   'install',
1 verbose cli   '--global',
1 verbose cli   'markmap-lib@0.11.1',
1 verbose cli   'netlify-cli@2.69.11',
1 verbose cli   '--color=always'
1 verbose cli ]
2 info using npm@6.14.8
3 info using node@v12.18.2
4 verbose npm-session 61b88cd81efdd033
5 silly install loadCurrentTree
6 silly install readGlobalPackageData
7 http fetch GET 304 https://registry.npmjs.org/netlify-cli 2556ms (from cache)
8 silly pacote version manifest for netlify-cli@2.69.11 fetched in 2573ms
9 http fetch GET 304 https://registry.npmjs.org/markmap-lib 3931ms (from cache)
10 silly pacote version manifest for markmap-lib@0.11.1 fetched in 3947ms
11 timing stage:loadCurrentTree Completed in 9038ms
12 silly install loadIdealTree
13 silly install cloneCurrentTreeToIdealTree
14 timing stage:loadIdealTree:cloneCurrentTree Completed in 68ms
15 silly install loadShrinkwrap
16 verbose stack RangeError: Maximum call stack size exceeded
16 verbose stack     at /usr/share/nodejs/aproba/index.js:39:16
16 verbose stack     at Array.forEach (<anonymous>)
16 verbose stack     at validate (/usr/share/nodejs/aproba/index.js:34:11)
16 verbose stack     at flatNameFromTree (/usr/share/npm/lib/install/flatten-tree.js:37:3)
16 verbose stack     at flatNameFromTree (/usr/share/npm/lib/install/flatten-tree.js:39:14)
16 verbose stack     at flatNameFromTree (/usr/share/npm/lib/install/flatten-tree.js:39:14)
16 verbose stack     at flatNameFromTree (/usr/share/npm/lib/install/flatten-tree.js:39:14)
16 verbose stack     at flatNameFromTree (/usr/share/npm/lib/install/flatten-tree.js:39:14)
16 verbose stack     at flatNameFromTree (/usr/share/npm/lib/install/flatten-tree.js:39:14)
16 verbose stack     at flatNameFromTree (/usr/share/npm/lib/install/flatten-tree.js:39:14)
16 verbose stack     at flatNameFromTree (/usr/share/npm/lib/install/flatten-tree.js:39:14)
16 verbose stack     at flatNameFromTree (/usr/share/npm/lib/install/flatten-tree.js:39:14)
16 verbose stack     at flatNameFromTree (/usr/share/npm/lib/install/flatten-tree.js:39:14)
16 verbose stack     at flatNameFromTree (/usr/share/npm/lib/install/flatten-tree.js:39:14)
16 verbose stack     at flatNameFromTree (/usr/share/npm/lib/install/flatten-tree.js:39:14)
16 verbose stack     at flatNameFromTree (/usr/share/npm/lib/install/flatten-tree.js:39:14)
17 verbose cwd /usr/local/lib/node_modules
18 verbose Linux 5.8.0-33-generic
19 verbose argv "/usr/bin/node" "/usr/bin/npm" "install" "--global" "markmap-lib@0.11.1" "netlify-cli@2.69.11" "--color=always"
20 verbose node v12.18.2
21 verbose npm  v6.14.8
22 error Maximum call stack size exceeded
23 verbose exit [ 1, true ]

```

<details>

**Comment by nlf** [comment]
do you, by chance, use something like nvm to manage your node versions for your user?

i ask because i see where you clearly installed npm@next-7 and got npm 7.3.0 when you ran `npm -v`, however the log from the error you pasted shows at the bottom that it's from npm 6.14.8. i'm wondering if your root user (since you ran the command with `sudo`) has a different version of node and/or npm.

what do you get if you run `sudo npm -v`?

**Comment by CyrusYip** [comment]
```
~ sudo npm -v 
7.3.0
~ npm -v 
7.3.0
~ sudo node -v
v12.18.2
~ node -v 
v12.18.2
```

I did not use any node version manager. At first, I installed [Node.js](https://packages.ubuntu.com/groovy/nodejs) 12.18.2 and [npm](https://packages.ubuntu.com/groovy/npm) 6.14.8 from Ubuntu software repo. Then I upgraded npm to 6.14.10 and 7.3.0 via `sudo npm install -g npm` and `sudo npm install -g npm@next-7`.

I discovered that I have two versions of npm on my computer and one version of Node.js.

```bash
~ where npm
/usr/local/bin/npm
/usr/bin/npm
/bin/npm
~ /usr/local/bin/npm -v
7.3.0
~ /usr/bin/npm -v
6.14.8
~ /bin/npm -v
6.14.8
~ ls -l /usr/local/bin/npm                                     
lrwxrwxrwx 1 root root 38 Dec 26 14:21 /usr/local/bin/npm -> ../lib/node_modules/npm/bin/npm-cli.js
~ ls -l /usr/bin/npm
lrwxrwxrwx 1 root root 27 Aug 25 15:10 /usr/bin/npm -> ../share/npm/bin/npm-cli.js
~ ls -l /bin/npm            
lrwxrwxrwx 1 root root 27 Aug 25 15:10 /bin/npm -> ../share/npm/bin/npm-cli.js
```

```bash
~ where node
/usr/bin/node
/bin/node
~ /usr/bin/node -v
v12.18.2
~ /bin/node -v    
v12.18.2
~ ls -l /usr/bin/node        
-rwxr-xr-x 1 root root 26616 Aug 20 02:47 /usr/bin/node
~ ls -l /bin/node             
-rwxr-xr-x 1 root root 26616 Aug 20 02:47 /bin/node
```

**Comment by nlf** [comment]
this sure makes it seem like `npm-check` is still running the wrong version of npm, i'm not sure what's going on with that since i'm unfamiliar with that tool.

that said, i can confirm that a plain `npm update -g` should work and currently does not for me either. i'll follow up again here when we have a fix!

**Comment by jelhouss** [comment]
I had a similar issue. I used a version of Node.js (forgot exactly, but I think it was **12.x**), however, I planned to upgrade to **v15.0.x** and then wanted to list my outdated packages. 
After I upgraded to **v15.0.1** (did not test with **v15.0.0** so perhaps this could be a particular problem to `15.0.1`, or generally `15.x`) and npm **v7.0.3**; `npm outdated` logged:

```
npm ERR! Only tag, version, and range are supported

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/user/.npm/_logs/2021-01-22T00_29_33_027Z-debug.log
```
Side note: I use `nvm` to manage versions, and the `node` bin and the `npm` bin that run when I used `outdated` was from `.nvm`.

So I decided to downgrade to a recent stable version, and I chose `v14.15.4` (packaged with npm `v6.14.10`), and after re-installing modules again, the `npm outdated` command actually listed the expected output and I updated my packages.

**Comment by ljharb** [comment]
@jelhouss on node 15, it'd be worth doing `npm install -g npm@7`, and trying with v7.4.3.

**Comment by nlf** [comment]
hi! this is fixed in 7.5.0, please let us know if you see it happen again!

**Comment by CyrusYip** [comment]
Thanks, @nlf . It worked!

```bash
~ sudo npm install -g npm@next-7                                   

changed 14 packages in 3s

11 packages are looking for funding
  run `npm fund` for details
~ npm -v
7.5.1
~ sudo npm update -g             
npm WARN deprecated har-validator@5.1.5: this library is no longer supported
npm WARN deprecated rollup-plugin-inject@3.0.2: This package has been deprecated and is no longer maintained. Please use @rollup/plugin-inject.
npm WARN deprecated request@2.88.2: request has been deprecated, see https://github.com/request/request/issues/3142
npm WARN deprecated urix@0.1.0: Please see https://github.com/lydell/urix#deprecated
npm WARN deprecated resolve-url@0.2.1: https://github.com/lydell/resolve-url#deprecated
npm WARN deprecated core-js@2.6.12: core-js@<3 is no longer maintained and not recommended for usage due to the number of issues. Please, upgrade your dependencies to the actual version of core-js@3.

added 592 packages, removed 4501 packages, and changed 1881 packages in 1m
```

My global packages are up-to-date now.

```bash
~ npm list -g | grep netlify-cli       
├─┬ netlify-cli@3.4.6
~ npm list -g | grep markmap-lib           
├─┬ markmap-lib@0.11.2
```

However, npm version was downgraded.

```bash
~ npm -v                    
6.14.11
```

Is the downgrade expected? Should I stick to npm 7? If so, how can I prevent the downgrade?
